data-node/src/main/java/org/graylog/datanode/bootstrap/preflight/DataNodePreflightGeneratePeriodical.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.datanode.bootstrap.preflight;
+
+import org.bouncycastle.operator.OperatorCreationException;
+import org.graylog.security.certutil.csr.CsrGenerator;
+import org.graylog.security.certutil.csr.exceptions.CSRGenerationException;
+import org.graylog.security.certutil.csr.storage.CsrMongoStorage;
+import org.graylog.security.certutil.csr.storage.CsrStorage;
+import org.graylog.security.certutil.privatekey.PrivateKeyEncryptedFileStorage;
+import org.graylog2.cluster.NodeNotFoundException;
+import org.graylog2.cluster.NodePreflightConfig;
+import org.graylog2.cluster.NodePreflightConfigService;
+import org.graylog2.cluster.NodeService;
+import org.graylog2.plugin.periodical.Periodical;
+import org.graylog2.plugin.system.NodeId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+
+@Singleton
+public class DataNodePreflightGeneratePeriodical extends Periodical {
+    private static final Logger LOG = LoggerFactory.getLogger(DataNodePreflightGeneratePeriodical.class);
+
+    private final NodePreflightConfigService nodePreflightConfigService;
+    private final NodeService nodeService;
+    private final NodeId nodeId;
+    private final PrivateKeyEncryptedFileStorage privateKeyEncryptedStorage;
+    private final CsrStorage csrStorage;
+
+    //TODO: decide on password handling
+    private static final String DEFAULT_PASSWORD = "admin";
+
+    @Inject
+    public DataNodePreflightGeneratePeriodical(final NodePreflightConfigService nodePreflightConfigService,
+                                               final NodeService nodeService,
+                                               final NodeId nodeId,
+                                               final CsrMongoStorage csrStorage) {
+        this.nodePreflightConfigService = nodePreflightConfigService;
+        this.nodeService = nodeService;
+        this.nodeId = nodeId;
+        this.csrStorage = csrStorage;
+        // TODO: merge with real storage
+        this.privateKeyEncryptedStorage = new PrivateKeyEncryptedFileStorage("privateKeyFilename.cert");
+    }
+
+    @Override
+    public void doRun() {
+        LOG.debug("checking if this DataNode is supposed to take configuration steps.");
+        var cfg = nodePreflightConfigService.getPreflightConfigFor(nodeId.getNodeId());
+        if (cfg == null) {
+            // write default config if none exists for this node
+            nodePreflightConfigService.save(NodePreflightConfig.builder().nodeId(nodeId.getNodeId()).state(NodePreflightConfig.State.UNCONFIGURED).build());
+        } else if (NodePreflightConfig.State.CONFIGURED.equals(cfg.state())) {
+            try {
+                var node = nodeService.byNodeId(nodeId);
```

Comment:
We need this invocation just to get a hostname... So maybe it should be done only once in this periodical, in the beginning?